File: dfs-bfs/bfs/[백준]2206. 벽 부수고 이동하기/solution.py
import sys
from collections import deque
input = sys.stdin.readline

dx = [1, -1, 0, 0]
dy = [0, 0, 1, -1]

n, m = map(int, input().split())
arr = [input() for _ in range(n)]
visited = [[0b00 for c in range(m)] for r in range(n)] # 00, 01, 10, 11
dq = deque()
dq.append((0, 0, 0, 1))
visited[0][0] = 0b11

def bfs():
    if n == m == 1: return 1
    while dq:
        x, y, w, d = dq.popleft()
        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]
            if nx < 0 or ny < 0 or nx >= n or ny >= m: continue
            if arr[nx][ny] == '0':
                if w == 0 and visited[nx][ny] & 0b01 == 0:
                    if nx == n - 1 and ny == m - 1: return d + 1
                    visited[nx][ny] |= 0b01
                    dq.append((nx, ny, 0, d + 1))
                if w == 1 and visited[nx][ny] & 0b10 == 0:
                    if nx == n - 1 and ny == m - 1: return d + 1
                    visited[nx][ny] |= 0b10
                    dq.append((nx, ny, 1, d + 1))
            elif arr[nx][ny] == '1' and w == 0:
                visited[nx][ny] |= 0b10
                dq.append((nx, ny, 1, d + 1))
    return -1

print(bfs())

File: dfs-bfs/bfs/[백준]2206. 벽 부수고 이동하기/solution.cpp
#include <bits/stdc++.h>
#define M 1000

using namespace std;

int dx[4] = {1, -1 ,0, 0};
int dy[4] = {0, 0, 1, -1};

int n, m;
int arr[M][M];
int visited[M][M];
queue<tuple<int, int, int, int>> q;

int bfs() {
    if (n == 1 && m == 1) return 1;
    visited[0][0] = 0b11;
    q.push({0, 0, 0, 1});
    while (!q.empty()) {
        int x, y, w, d;
        tie(x, y, w, d) = q.front(); q.pop();
        for (int i = 0; i < 4; i++) {
            int nx = x + dx[i];
            int ny = y + dy[i];
            if (nx < 0 || ny < 0 || nx >= n || ny >= m) continue;
            if (arr[nx][ny] == 0) {
                if ((visited[nx][ny] & (1 << w)) != 0) continue;
                if (nx == n - 1 && ny == m - 1) return d + 1;
                visited[nx][ny] |= (1 << w);
                q.push({nx, ny, w, d + 1}); 
            }
            else if (w == 0) {
                visited[nx][ny] |= 0b10;
                q.push({nx, ny, 1, d + 1});
            }
        }
    }
    return -1;
}

int main() {
    ios::sync_with_stdio(false); cin.tie(nullptr);
    cin >> n >> m;
    for (int i = 0; i < n; i++) {
        string s;
        cin >> s;
        for (int j = 0; j < m; j++)
            arr[i][j] = s[j] - '0';
    }
    cout << bfs();
}